tools/@aws-cdk/lazify/lib/index.ts

```diff
@@ -0,0 +1,220 @@
+ *
+ * That is:
+ *
+ * - Find all top-level require() assignments, and replace them with a function that performs
+ *   the require(). That way, the require() is only done if any of the objects from its scope
+ *   are actually used.
+ * - Find all (transpiled) `export * from 'xyz';` statements (by searching for an invocation
+ *   of `__exportStar()`): load the actual module, enumerate the entries, and create a getter
+ *   for each entry.
+ */
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import * as ts from 'typescript';
+
+type LogFn = (...x: string[]) => void;
+
+export async function transformFile(filename: string, verbose: boolean) {
+  progress(filename, '... ');
+  const contents = await fs.readFile(filename, { encoding: 'utf-8' });
+  const transformed = transformFileContents(filename, contents, progress);
+  await fs.writeFile(filename, transformed, { encoding: 'utf-8' });
+  progress(' Done!\n');
+
+  function progress(...x: string[]) {
+    if (verbose) {
+      process.stderr.write(x.join(' '));
+    }
+  }
+}
+
+export function transformFileContents(filename: string, contents: string, progress?: LogFn) {
+  const sourceFile = ts.createSourceFile(
+    filename,
+    contents,
+    ts.ScriptTarget.Latest,
+    true // setParentNodes, need this for tree analysis
+  );
+
+  // Find all top-level requires and turn them into a function
+  const topLevelAssignments = sourceFile.statements
+    .filter(ts.isVariableStatement)
+    .filter((stmt) => stmt.declarationList.declarations.length === 1)
+    .map((stmt) => [stmt, stmt.declarationList.declarations[0]] as const);
+
+  progress?.(`${topLevelAssignments.length} declarations`, '... ');
+
+  const topLevelRequires = topLevelAssignments
+    .flatMap(([stmt, a]) => a.initializer && ts.isCallExpression(a.initializer)
+      && ts.isIdentifier(a.initializer.expression) && a.initializer.expression.text === 'require'
+      && ts.isStringLiteral(a.initializer.arguments[0])
+      && ts.isIdentifier(a.name)
+      ? [[stmt, a.name, a.initializer.arguments[0].text] as const] : []);
+
+  progress?.(`${topLevelRequires.length} requires`, '... ');
+
+  let file = sourceFile;
+
+  for (const [stmt, binding, moduleName] of topLevelRequires) {
+    const result = ts.transform(file, [(ctx: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
+      const factory = ctx.factory;
+      const visit: ts.Visitor = node => {
+        // If this is the statement, replace it with a function definition
+        if (node === stmt) {
+          return createVariable(factory, binding,
+            factory.createArrowFunction(undefined, undefined, [], undefined, undefined,
+              factory.createBlock([
+                // tmp = require(...)
+                createVariable(factory, 'tmp', factory.createCallExpression(factory.createIdentifier('require'), [], [factory.createStringLiteral(moduleName)])),
+
+                // <this_fn> = () => tmp
```

Comment:
Consider commenting on why we need this. My naive expectation would be to have something like:

```ts
var x = () => { return require("x"); };
```

without any need for a temporary variable.